File: packages/kits/AriaMcpKit/Tests/AriaMCPTests/AriaV2EstateDiagnosticsTests.swift
import Foundation
import Testing
@testable import AriaMCP

@Suite("ARIA v2 typed estate diagnostics")
struct AriaV2EstateDiagnosticsTests {
    private let estateID = UUID(uuidString: "aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa")!

    @Test("All diagnostics accept only the shared optional estate_id")
    func strictRequests() async throws {
        let diagnostics = service(provider: FakeEstateDiagnosticsProvider(estateID: estateID))
        for operation in [
            diagnostics.ping, diagnostics.status, diagnostics.map,
            diagnostics.drainStatus, diagnostics.rebuildStatus, diagnostics.timingReport,
        ] {
            _ = try await operation(.object(["estate_id": .string(estateID.uuidString.uppercased())]))
            await #expect(throws: JSONRPCError.self) {
                _ = try await operation(.object(["unknown": .bool(true)]))
            }
            await #expect(throws: JSONRPCError.self) {
                _ = try await operation(.object(["estate_id": .string("not-a-uuid")]))
            }
        }
    }

    @Test("Each service projects its operation-specific typed data")
    func typedDataProjection() async throws {
        let provider = FakeEstateDiagnosticsProvider(estateID: estateID)
        let diagnostics = service(provider: provider)

        let ping = try await diagnostics.ping(arguments: .object([:]))
        let status = try await diagnostics.status(arguments: .object([:]))
        let map = try await diagnostics.map(arguments: .object([:]))
        let drains = try await diagnostics.drainStatus(arguments: .object([:]))
        let rebuild = try await diagnostics.rebuildStatus(arguments: .object([:]))
        let timing = try await diagnostics.timingReport(arguments: .object([:] ))

        #expect((data(ping).map { Set($0.keys) } ?? Set<String>()) == Set(["estate_id", "estate_name", "state", "build_serial"]))
        #expect(data(ping)?["state"] == .string("mounted"))
        #expect(data(ping)?["build_serial"] == .string("fixture-build"))
        // recall_trace_count is present because the fixture supplies a count;
        // shared_content_migration is absent because the fixture has no
        // migration record, which is the shape an estate that never ran
        // detection returns.
        #expect((data(status).map { Set($0.keys) } ?? Set<String>()) == Set([
            "estate_id", "estate_name", "memory_count", "fact_count", "drains",
            "fdc_recalculation", "recall_trace_count", "sync_state",
            "subjects_bearing", "subjects_eligible",
        ]))
        #expect(data(status)?["memory_count"] == .integer(2))
        #expect(data(map)?["wings"]?.arrayValue?.first?.objectValue?["rooms"]?.arrayValue?.count == 1)
        #expect((data(drains).map { Set($0.keys) } ?? Set<String>()) == Set(["drains"]))
        #expect((data(drains)?["drains"]?.arrayValue?.first?.objectValue.map { Set($0.keys) } ?? Set<String>()) == Set(["name", "state", "pending"]))
        #expect(data(drains)?["drains"]?.arrayValue?.first?.objectValue?["pending"] == .integer(3))
        #expect(data(rebuild)?["state"] == .string("running"))
        #expect((data(rebuild).map { Set($0.keys) } ?? Set<String>()) == Set(["state"]))
        #expect((data(timing).map { Set($0.keys) } ?? Set<String>()) == Set(["since_ms", "watermark_ms", "truncated"]))
        #expect(data(timing)?["since_ms"] == .integer(0))
        #expect(data(timing)?["watermark_ms"] == .integer(1_700_000_001_000))
        #expect(meta(status)?["effect"] == .string("read"))
        #expect(meta(status)?["session_id"] == .string("test-session"))
        #expect(meta(status)?["observed_at"] == .string("2023-11-14T22:13:20Z"))
        #expect(await provider.calls == [.estatePing, .estateStatus, .estateMap, .drainStatus, .rebuildStatus, .timingReport])
    }

    @Test("Access refusal prevents a lower provider call")
    func accessGateRunsBeforeProvider() async throws {
        let provider = FakeEstateDiagnosticsProvider(estateID: estateID)
        let refusal = AriaV2OperationalRefusal(
            code: "estate_access_denied", message: "Session may not inspect this estate.", retryable: false)
        let diagnostics = service(provider: provider, gate: RefusingGate(refusal: refusal))
        let response = try await diagnostics.status(arguments: .object([:]))

        #expect(response.objectValue?["isError"] == .bool(true))
        #expect(response.objectValue?["structuredContent"]?.objectValue?["error"]?.objectValue?["code"] == .string("estate_access_denied"))
        #expect(await provider.calls.isEmpty)
    }

    @Test("Unavailable ping is an operational refusal")
    func unavailablePingIsRefusal() async throws {
        let provider = FakeEstateDiagnosticsProvider(
            estateID: estateID,
            pingRefusal: .init(
                code: "estate_unavailable",
                message: "The selected estate is quiesced and not accepting new work.",
                retryable: false))
        let response = try await service(provider: provider).ping(arguments: .object([:]))
        #expect(response.objectValue?["isError"] == .bool(true))
        #expect(response.objectValue?["structuredContent"]?.objectValue?["error"]?.objectValue?["code"] == .string("estate_unavailable"))
    }

    @Test("A different selected estate fails before access or collection")
    func selectedEstateMustMatchContext() async throws {
        let provider = FakeEstateDiagnosticsProvider(estateID: estateID)
        let diagnostics = service(provider: provider)
        await #expect(throws: JSONRPCError.self) {
            _ = try await diagnostics.ping(arguments: .object([
                "estate_id": .string(UUID().uuidString),
            ]))
        }
        #expect(await provider.calls.isEmpty)
    }

    private func service(
        provider: FakeEstateDiagnosticsProvider,
        gate: any AriaV2EstateDiagnosticsAccessGate = AriaV2AllowEstateDiagnosticsAccess()
    ) -> AriaV2EstateDiagnostics {
        AriaV2EstateDiagnostics(
            provider: provider,
            context: .init(
                estateID: estateID,
                estateName: "Fixture Estate",
                callerID: "fixture-caller",
                serverIdentity: "fixture-server",
                sessionID: "test-session",
                buildSerial: "fixture-build",
                now: { Date(timeIntervalSince1970: 1_700_000_000) },
                accessGate: gate))
    }

    private func data(_ response: JSONValue) -> [String: JSONValue]? {
        response.objectValue?["structuredContent"]?.objectValue?["data"]?.objectValue
    }

    private func meta(_ response: JSONValue) -> [String: JSONValue]? {
        response.objectValue?["structuredContent"]?.objectValue?["meta"]?.objectValue
    }
}

private actor FakeEstateDiagnosticsProvider: AriaV2EstateDiagnosticsProvider {
    let estateID: UUID
    let pingRefusal: AriaV2OperationalRefusal?
    private(set) var calls: [AriaV2EstateDiagnosticOperation] = []

    init(estateID: UUID, pingRefusal: AriaV2OperationalRefusal? = nil) {
        self.estateID = estateID
        self.pingRefusal = pingRefusal
    }

    func ping(context: AriaV2EstateDiagnosticsContext) async throws -> AriaV2EstatePingResult {
        calls.append(.estatePing)
        if let pingRefusal { return .refusal(pingRefusal) }
        return .mounted(.init(
            estateID: estateID,
            estateName: context.estateName,
            state: "mounted",
            buildSerial: context.buildSerial))
    }

    func status(context: AriaV2EstateDiagnosticsContext) async throws -> AriaV2EstateStatusData {
        calls.append(.estateStatus)
        return .init(
            estateID: estateID,
            estateName: context.estateName,
            memoryCount: 2,
            factCount: 4,
            drains: [.init(name: "corpus_encode", state: "draining", pending: 3)],
            fdcRecalculation: "missing",
            recallTraceCount: 7,
            syncState: "local-only",
            subjectsBearing: 1,
            subjectsEligible: 2,
            sharedContentMigration: nil)
    }

    func map(context: AriaV2EstateDiagnosticsContext) async throws -> AriaV2EstateMapData {
        _ = context
        calls.append(.estateMap)
        return .init(estateID: estateID, wings: [.init(name: "Memory", rooms: [.init(name: "Planning", memoryCount: 3)])])
    }

    func drains(context: AriaV2EstateDiagnosticsContext) async throws -> AriaV2DrainStatusData {
        _ = context
        calls.append(.drainStatus)
        return .init(drains: [.init(name: "corpus_encode", state: "draining", pending: 3)])
    }

    func rebuild(context: AriaV2EstateDiagnosticsContext) async throws -> AriaV2RebuildStatusData {
        _ = context
        calls.append(.rebuildStatus)
        return .init(state: "running")
    }

    func timing(context: AriaV2EstateDiagnosticsContext) async throws -> AriaV2TimingReportData {
        _ = context
        calls.append(.timingReport)
        return .init(sinceMilliseconds: 0, watermarkMilliseconds: 1_700_000_001_000, truncated: false)
    }
}

private struct RefusingGate: AriaV2EstateDiagnosticsAccessGate {
    let refusal: AriaV2OperationalRefusal

    func admit(
        _ operation: AriaV2EstateDiagnosticOperation,
        context: AriaV2EstateDiagnosticsContext
    ) async -> AriaV2OperationalRefusal? {
        _ = operation
        _ = context
        return refusal
    }
}

File: packages/kits/AriaMcpKit/rust/tests/aria_v2_estate_diagnostics_tests.rs

#[path = "../src/v2/estate_diagnostics.rs"]
mod estate_diagnostics;

use estate_diagnostics::{
    DiagnosticsFact, DiagnosticsLifecycle, DiagnosticsMemory, EstateDiagnosticsAuthority,
    EstateDiagnosticsContext, EstateDiagnosticsFailure, EstateDiagnosticsGrant, EstateDiagnosticsOperation,
    EstateDiagnosticsRequest, EstateDiagnosticsService, EstateDiagnosticsSnapshot, EstateDrain,
    EstateDrainState, EstateRebuildState, EstateTiming,
};
use serde_json::json;
use std::sync::{Arc, atomic::{AtomicUsize, Ordering}};
use uuid::Uuid;

fn estate_id() -> Uuid { Uuid::parse_str("aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa").unwrap() }

fn context() -> EstateDiagnosticsContext {
    EstateDiagnosticsContext {
        caller_binding: "caller-a".to_owned(),
        session_id: "session-a".to_owned(),
        clock_millis: 1_700_000_000_000,
        build_serial: "build-20260908".to_owned(),
    }
}

fn request() -> EstateDiagnosticsRequest { EstateDiagnosticsRequest { estate_id: Some(estate_id()) } }

fn snapshot(mounted: bool) -> EstateDiagnosticsSnapshot {
    EstateDiagnosticsSnapshot {
        estate_id: estate_id(),
        estate_name: "Test Estate".to_owned(),
        mounted,
        memories: vec![
            DiagnosticsMemory { wing: "Agents".to_owned(), room: "inbox".to_owned(), lifecycle: DiagnosticsLifecycle::CurrentClusterA, bulk_exportable: true },
            DiagnosticsMemory { wing: "Agents".to_owned(), room: "inbox".to_owned(), lifecycle: DiagnosticsLifecycle::CurrentClusterA, bulk_exportable: true },
            DiagnosticsMemory { wing: "Private".to_owned(), room: "secrets".to_owned(), lifecycle: DiagnosticsLifecycle::CurrentClusterA, bulk_exportable: false },
            DiagnosticsMemory { wing: "Archived".to_owned(), room: "old".to_owned(), lifecycle: DiagnosticsLifecycle::Other, bulk_exportable: true },
        ],
        facts: vec![
            DiagnosticsFact { lifecycle: DiagnosticsLifecycle::CurrentClusterA, bulk_exportable: true },
            DiagnosticsFact { lifecycle: DiagnosticsLifecycle::CurrentClusterA, bulk_exportable: false },
            DiagnosticsFact { lifecycle: DiagnosticsLifecycle::Other, bulk_exportable: true },
        ],
        drains: vec![EstateDrain { name: "corpus_encode".to_owned(), state: EstateDrainState::Draining, pending: 4 }],
        rebuild: EstateRebuildState::Running,
        timing: EstateTiming { watermark_ms: 1_700_000_123_456, truncated: true },
        recall_trace_count: Some(7),
        sync_state: "local-only".to_owned(),
        subjects_bearing: 1,
        subjects_eligible: 2,
        shared_content_migration: None,
        fdc_floor: None,
    }
}

struct Authority {
    snapshot: EstateDiagnosticsSnapshot,
    expected_context: EstateDiagnosticsContext,
    heavy_inventory_reads: Arc<AtomicUsize>,
}

impl Authority {
    fn new(snapshot: EstateDiagnosticsSnapshot) -> Self {
        Self::with_counter(snapshot, Arc::new(AtomicUsize::new(0)))
    }

    fn with_counter(snapshot: EstateDiagnosticsSnapshot, heavy_inventory_reads: Arc<AtomicUsize>) -> Self {
        Self { snapshot, expected_context: context(), heavy_inventory_reads }
    }

    fn verify(&self, context: &EstateDiagnosticsContext) -> Result<(), EstateDiagnosticsFailure> {
        if context != &self.expected_context {
            return Err(EstateDiagnosticsFailure::operational("access_denied", "Context binding did not match.", false));
        }
        Ok(())
    }
}

impl EstateDiagnosticsAuthority for Authority {
    fn authorize(
        &self,
        operation: EstateDiagnosticsOperation,
        requested_estate_id: Option<Uuid>,
        context: &EstateDiagnosticsContext,
    ) -> Result<EstateDiagnosticsGrant, EstateDiagnosticsFailure> {
        self.verify(context)?;
        if requested_estate_id != Some(estate_id()) {
            return Err(EstateDiagnosticsFailure::operational("access_denied", "Estate is not authorized.", false));
        }
        let _ = operation;
        Ok(EstateDiagnosticsGrant { estate_id: estate_id(), estate_name: "Test Estate".to_owned() })
    }

    fn snapshot(
        &self,
        operation: EstateDiagnosticsOperation,
        _: &EstateDiagnosticsGrant,
        context: &EstateDiagnosticsContext,
    ) -> Result<EstateDiagnosticsSnapshot, EstateDiagnosticsFailure> {
        self.verify(context)?;
        if operation != EstateDiagnosticsOperation::Ping {
            self.heavy_inventory_reads.fetch_add(1, Ordering::SeqCst);
        }
        Ok(self.snapshot.clone())
    }

    fn revalidate(
        &self,
        _: EstateDiagnosticsOperation,
        _: &EstateDiagnosticsGrant,
        context: &EstateDiagnosticsContext,
    ) -> Result<(), EstateDiagnosticsFailure> {
        self.verify(context)
    }
}

#[test]
fn request_accepts_only_optional_estate_id() {
    let decoded = EstateDiagnosticsRequest::decode(&json!({"estate_id": estate_id().to_string()})).unwrap();
    assert_eq!(decoded.estate_id, Some(estate_id()));
    assert!(EstateDiagnosticsRequest::decode(&json!({"since_ms": 0})).is_err());
    assert!(EstateDiagnosticsRequest::decode(&json!({"estate_id": 7})).is_err());
}

#[test]
fn status_and_map_apply_cluster_a_and_public_bulk_ceiling() {
    let service = EstateDiagnosticsService::new(Authority::new(snapshot(true)));
    let status = service.status(request(), &context()).unwrap();
    assert_eq!(status.estate_id, "aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa");
    assert_eq!(status.estate_name, "Test Estate");
    assert_eq!(status.memory_count, 2);
    assert_eq!(status.fact_count, 1);
    assert_eq!(status.drains.len(), 1);
    assert_eq!(status.fdc_recalculation, "missing");
    assert_eq!(serde_json::to_value(&status).unwrap(), json!({
        "estate_id": "aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa",
        "estate_name": "Test Estate",
        "memory_count": 2,
        "fact_count": 1,
        "fdc_recalculation": "missing",
        "drains": [{"name": "corpus_encode", "state": "draining", "pending": 4}],
        // recall_trace_count is present because the fixture supplies one.
        // shared_content_migration is ABSENT, which is the shape an estate
        // that never ran detection returns — omitted, not null.
        "recall_trace_count": 7,
        "sync_state": "local-only",
        "subjects_bearing": 1,
        "subjects_eligible": 2,
    }));

    let map = service.map(request(), &context()).unwrap();
    assert_eq!(map.wings.len(), 1);
    assert_eq!(map.wings[0].name, "Agents");
    assert_eq!(map.wings[0].rooms[0].name, "inbox");
    assert_eq!(map.wings[0].rooms[0].memory_count, 2);
    assert_eq!(serde_json::to_value(&map).unwrap(), json!({
        "estate_id": "aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa",
        "wings": [{"name": "Agents", "rooms": [{"name": "inbox", "memory_count": 2}]}],
    }));
}

#[test]
fn ping_refuses_nonmounted_and_binds_build_context() {
    let heavy_inventory_reads = Arc::new(AtomicUsize::new(0));
    let authority = Authority::with_counter(snapshot(true), Arc::clone(&heavy_inventory_reads));
    let live = EstateDiagnosticsService::new(authority);
    let ping = live.ping(request(), &context()).unwrap();
    assert_eq!(ping.state, "mounted");
    assert_eq!(ping.build_serial, "build-20260908");
    assert_eq!(heavy_inventory_reads.load(Ordering::SeqCst), 0,
        "ping must not request an inventory snapshot");
    assert_eq!(serde_json::to_value(&ping).unwrap(), json!({
        "estate_id": "aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa",
        "estate_name": "Test Estate",
        "state": "mounted",
        "build_serial": "build-20260908",
    }));

    let offline = EstateDiagnosticsService::new(Authority::new(snapshot(false)));
    let error = offline.ping(request(), &context()).unwrap_err();
    assert_eq!(error.code, "estate_unavailable");

    let mut wrong_context = context();
    wrong_context.session_id = "other-session".to_owned();
    let error = live.status(request(), &wrong_context).unwrap_err();
    assert_eq!(error.code, "access_denied");
}

#[test]
fn snapshot_seam_receives_the_operation_before_data_collection() {
    let heavy_inventory_reads = Arc::new(AtomicUsize::new(0));
    let authority = Authority::with_counter(snapshot(true), Arc::clone(&heavy_inventory_reads));
    let service = EstateDiagnosticsService::new(authority);
    service.status(request(), &context()).unwrap();
    assert_eq!(heavy_inventory_reads.load(Ordering::SeqCst), 1);
}

#[test]
fn drain_rebuild_and_timing_use_the_frozen_data_shapes() {
    let service = EstateDiagnosticsService::new(Authority::new(snapshot(true)));
    let drains = service.drain(request(), &context()).unwrap();
    assert_eq!(drains.drains, vec![EstateDrain {
        name: "corpus_encode".to_owned(), state: EstateDrainState::Draining, pending: 4,
    }]);
    let rebuild = service.rebuild(request(), &context()).unwrap();
    assert_eq!(rebuild.state, EstateRebuildState::Running);
    let timing = service.timing(request(), &context()).unwrap();
    assert_eq!(timing.since_ms, 0);
    assert_eq!(timing.watermark_ms, 1_700_000_123_456);
    assert!(timing.truncated);
    assert_eq!(serde_json::to_value(&drains).unwrap(), json!({
        "drains": [{"name": "corpus_encode", "state": "draining", "pending": 4}],
    }));
    assert_eq!(serde_json::to_value(&rebuild).unwrap(), json!({"state": "running"}));
    assert_eq!(serde_json::to_value(&timing).unwrap(), json!({
        "since_ms": 0,
        "watermark_ms": 1_700_000_123_456i64,
        "truncated": true,
    }));
}
